File: project/strategy/models/antifraud/AfBase.php
<?php

namespace app\models\antifraud;

use Yii;

/**
 * This is the model class for table "af_base".
 *
 * @property string $id
 * @property string $request_id
 * @property integer $aid
 * @property string $user_id
 * @property string $loan_id
 * @property string $jxlstat_id
 * @property string $create_time
 */
class AfBase extends BaseDBModel
{
    /**
     * @inheritdoc
     */
    public static function tableName()
    {
        return 'af_base';
    }

    /**
     * @inheritdoc
     */
    public function rules()
    {
        return [
            [['request_id', 'aid', 'user_id', 'loan_id', 'jxlstat_id'], 'integer'],
            [['create_time'], 'required'],
            [['create_time', 'modify_time'], 'safe']
        ];
    }

    /**
     * @inheritdoc
     */
    public function attributeLabels()
    {
        return [
            'id' => 'ID',
            'request_id' => '请求处理id',
            'aid' => '业务ID',
            'user_id' => '用户ID',
            'loan_id' => '贷款ID',
            'jxlstat_id' => '聚信立stat表的id',
            'create_time' => '创建时间',
            'modify_time' => '修改时间',
        ];
    }

    public function getBase($where) {
        return $this->find()->where($where)->orderby('ID DESC')->limit(1)->one();
    }

    public function addBase($baseInfo) {
        $error = $this->chkAttributes($baseInfo);
        if ($error) {
            return $this->returnError(null, current($error));
        }
        //3 保存数据
        $result = $this->save();
        if (!$result) {
            return $this->returnError(null, '保存失败');
        } else {
            return $result;
        }
    }
}

File: project/pyalg_api/model/antifraud/af_base.py
# -*- coding: utf-8 -*-
import os
from lib.application import db
from lib.logger import logger
from .base import Base
from datetime import datetime, timedelta
from sqlalchemy import and_

class AfBase(db.Model, Base):
    # 指定数据库
    __bind_key__ = 'xhh_antifraud'
    __tablename__ = 'af_base'

    id = db.Column(db.BigInteger, primary_key=True)
    request_id = db.Column(db.BigInteger, nullable=False, index=True)
    aid = db.Column(db.Integer,nullable=False)
    user_id = db.Column(db.BigInteger, nullable=False, index=True)
    loan_id = db.Column(db.BigInteger, nullable=False, index=True)
    jxlstat_id = db.Column(db.BigInteger, nullable=False, index=True)
    match_status = db.Column(db.Integer, nullable=False)
    create_time = db.Column(db.DateTime, nullable=False)
    modify_time = db.Column(db.DateTime, nullable=False)

    def getById(self, id):
        db_base = self.query.get(int(id))
        return db_base

    def addBaseData(self, dict_data):
        try:
            if dict_data is None or len(dict_data) == 0:
                return False
            return self.addJacByDict(dict_data)
        except Exception as e:
            logger.error(self.__class__.__name__ + " addOne:%s" % e)
            return 0

    def addJacByDict(self, dict_data):
        for k, v in list(dict_data.items()):
            setattr(self, k, v)
        return self.save()

    def save(self):
        db.session.add(self)
        try:
            db.session.flush()
            db.session.commit()
            db.session.refresh(self)
            return self.id
        except Exception:
            db.session.rollback()
            return 0

    def getMulMatchData(self, start_time):
        '''
        获取需要匹配二级关系的数据
        '''
        where = and_(
            AfBase.create_time >= start_time,
            AfBase.match_status == 2
        )

        res = db.session.query(AfBase)   \
            .filter(where) \
            .order_by(AfBase.create_time) \
            .limit(500) \
            .all()

        return res

    def getJaccardData(self, start_time):
        '''
        获取需要匹配间接关系(jaccard)的数据
        '''
        where = and_(
            AfBase.create_time >= start_time,
            AfBase.match_status == 4
        )

        res = db.session.query(AfBase) \
            .filter(where) \
            .order_by(AfBase.create_time) \
            .limit(500) \
            .all()

        return res
    def getRelationData(self, start_time):
        '''
        获取已存在关系(relation)的数据
        '''
        where = and_(
            AfBase.create_time >= start_time,
            AfBase.match_status == 6
        )

        res = db.session.query(AfBase) \
            .filter(where) \
            .order_by(AfBase.create_time) \
            .limit(500) \
            .all()

        return res
    def getJaccardDataByids(self, ids):
        '''
        获取需要匹配间接关系(jaccard)的数据
        '''
        where = and_(
            AfBase.match_status == 4
        )
        res = db.session.query(AfBase) \
            .filter(AfBase.match_status.in_([4,5,6])) \
            .order_by(AfBase.create_time) \
            .group_by(AfBase.user_id) \
            .filter(AfBase.user_id.in_(ids)) \
            .limit(500) \
            .all()
        return res

    def getAfBaseData(self, start_time):
        where = and_(
            AfBase.create_time >= start_time, 
            AfBase.match_status == 0, #初始
            AfBase.aid.in_([1,14])
        )

        res = db.session.query(AfBase)   \
            .filter(where) \
            .order_by(AfBase.create_time) \
            .limit(1000) \
            .all()

        return res

    def lockMatchStatus(self, data, status):
        if len(data) == 0:
            return False

        now = datetime.now()
        for o in data:
            o.modify_time = now
            o.match_status = status  # 锁定 | 完成

        db.session.commit()

        return True

    def finishMatched(self, match_status):
        '''
        结束二级关系匹配流程
        '''
        self.match_status = match_status
        self.modify_time = datetime.now()
        db.session.commit()

    #
    def getBaseForbaseid(self, base_id):
        if not base_id:
            return False
        return db.session.query(AfBase).filter(AfBase.id==base_id).first()

    